pkg/batcher/controllers/main_controller.go

```diff
@@ -0,0 +1,281 @@
+	if err != nil {
+		errStr = fmt.Sprintf("NewRequest create fail: %v", err)
+		return &errStr
+	}
+	req.Header.Add("Content-Type", batcherInfo.ContentType)
+	defer req.Body.Close()
+	client := &http.Client{Timeout: batcherInfo.Timeout}
+	resp, err := client.Do(req)
+	if err != nil {
+		errStr = fmt.Sprintf("NewRequest send fail: %v", err)
+		return &errStr
+	}
+	defer resp.Body.Close()
+	result, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		errStr = fmt.Sprintf("Response read fail: %v", err)
+		return &errStr
+	}
+	err = json.Unmarshal(result, &batcherInfo.Predictions)
+	if err != nil {
+		errStr = fmt.Sprintf("Response unmarshal fail: %v, %s", err, string(result))
+		return &errStr
+	} else {
+		log.Info("CallService", "Results", string(result))
+	}
+	return nil
+}
+
+func (batcherInfo *BatcherInfo) InitializeInfo() {
+	batcherInfo.BatchID = ""
+	batcherInfo.CurrentInputLen = 0
+	batcherInfo.Instances = make([]interface{}, 0)
+	batcherInfo.Predictions = Predictions{}
+	batcherInfo.Info = make(map[*context.Context] InputInfo)
+	batcherInfo.Start = GetNowTime()
+	batcherInfo.Now = batcherInfo.Start
+}
+
+func (batcherInfo *BatcherInfo) BatchPredict() {
+	err := batcherInfo.CallService()
+	if err != nil {
+		log.Error(errors.New(*err), "")
+		for _, v := range batcherInfo.Info {
+			res := Response{
+				Message: *err,
+				BatchID: "",
+				Predictions: nil,
+			}
+			*v.ChannelOut <- res
+		}
+	} else {
+		batcherInfo.BatchID = GenerateUUID()
+		for _, v := range batcherInfo.Info {
+			predictions := make([]interface{}, 0)
+			for _, index := range v.Index {
+				predictions = append(predictions, batcherInfo.Predictions.Predictions[index])
+			}
+			res := Response{
+				Message: "",
+				BatchID: batcherInfo.BatchID,
+				Predictions: predictions,
+			}
+			if jsonStr, err := json.Marshal(res); err == nil {
+				log.Info("BatchPredict", "Results", string(jsonStr))
+			} else {
+				log.Error(errors.New("marshal response fail"), "")
+			}
+			*v.ChannelOut <- res
+		}
+	}
+	batcherInfo.InitializeInfo()
+}
+
+func (batcherInfo *BatcherInfo) Batcher() {
+	for {
+		select {
+		case req := <- channelIn:
+			if len(batcherInfo.Instances) == 0 {
+				batcherInfo.Start = GetNowTime()
+			}
+			batcherInfo.CurrentInputLen = len(batcherInfo.Instances)
+			batcherInfo.Instances = append(batcherInfo.Instances, *req.Instances...)
+			var index = make([]int, 0)
+			for i := 0; i < len(*req.Instances); i++ {
+				index = append(index, batcherInfo.CurrentInputLen + i)
+			}
+			batcherInfo.Info[req.ContextInput] = InputInfo{
+				req.ChannelOut,
+				index,
+			}
+			batcherInfo.CurrentInputLen = len(batcherInfo.Instances)
+		case <- time.After(SleepTime):
+		}
+		batcherInfo.Now = GetNowTime()
+		if batcherInfo.CurrentInputLen >= batcherInfo.MaxBatchSize ||
+			(float64(batcherInfo.Now.Sub(batcherInfo.Start).Milliseconds()) >= batcherInfo.MaxLatency &&
+				batcherInfo.CurrentInputLen > 0) {
+			batcherInfo.BatchPredict()
```

Comment:
just came across https://docs.bentoml.org/en/latest/guides/micro_batching.html, I think according to the logic here we will have to wait for previous batch to finish before making next one. I am fine with this as an initial implementation, but let's make an issue to follow up.